libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Decoding/Decoder.hs

```diff
@@ -861,49 +859,84 @@ decodeSetEnforceNoDuplicates ::
+  (b -> Decoder s a) ->
+  Decoder s (Int, b)
+decodeListLikeWithCount decodeLenOrIndef insert decodeElement = do
+  decodeLenOrIndef >>= \case
+    Just len -> loop (\x -> pure (x >= len)) 0 mempty
+    Nothing -> loop (\_ -> decodeBreakOr) 0 mempty
+  where
+    loop :: (Int -> Decoder s Bool) -> Int -> b -> Decoder s (Int, b)
+    loop condition = go
+      where
+        go !count !acc = do
+          shouldStop <- condition count
+          if shouldStop
+            then pure (count, acc)
+            else do
+              element <- decodeElement acc
+              go (count + 1) (insert element acc)
+    {-# INLINE loop #-}
+{-# INLINE decodeListLikeWithCount #-}
+
+-- | Decode a collection of values with ability to supply length decoder. Duplicates are not
 -- allowed.
 decodeListLikeEnforceNoDuplicates ::
-  forall s a f.
-  Monoid (f a) =>
-  -- | Check for membership. Decoder will fail if this predicate returns True
-  (a -> f a -> Bool) ->
+  forall s a b c.
+  Monoid b =>
+  Decoder s (Maybe Int) ->
   -- | Add an element into the decoded List like data structure
-  (a -> f a -> f a) ->
+  (a -> b -> b) ->
+  -- | Get the final data structure and the number of elements it has.
+  (b -> (Int, c)) ->
   Decoder s a ->
-  Decoder s (f a)
-decodeListLikeEnforceNoDuplicates isMember insert decodeElement = do
-  decodeListLenOrIndef >>= \case
-    Just len -> loop (\x -> pure (x - 1, x <= 0)) len mempty
-    Nothing -> loop (\x -> (,) x <$> decodeBreakOr) () mempty
-  where
-    loop :: (t -> Decoder s (t, Bool)) -> t -> f a -> Decoder s (f a)
-    loop condition prevStep !acc = do
-      (nextStep, shouldStop) <- condition prevStep
-      if shouldStop
-        then pure acc
-        else do
-          a <- decodeElement
-          when (a `isMember` acc) $ fail "Duplicate key detected in the List"
-          loop condition nextStep (insert a acc)
+  Decoder s c
+decodeListLikeEnforceNoDuplicates decodeLenOrIndef insert getFinalWithCount decodeElement = do
+  (count, result) <- decodeListLikeWithCount decodeLenOrIndef insert (const decodeElement)
+  let (len, finalResult) = getFinalWithCount result
+  when (len /= count) $
+    fail $
+      "Final number of elements: "
+        <> show len
+        <> " does not match the total count that was decoded: "
+        <> show count
+  pure finalResult
 {-# INLINE decodeListLikeEnforceNoDuplicates #-}
 
 -- | Decode a Set as a either a definite or indefinite list. Duplicates are not
 -- allowed. Set tag 258 is permitted, but not enforced.
 decodeSetLikeEnforceNoDuplicates ::
-  forall s a f.
-  Monoid (f a) =>
-  -- | Check for membership. Decoder will fail if this predicate returns True
-  (a -> f a -> Bool) ->
-  -- | Add an aelement into the decoded Set like data structure
-  (a -> f a -> f a) ->
+  forall s a b c.
+  Monoid b =>
+  -- | Add an element into the decoded Set like data structure
+  (a -> b -> b) ->
+  -- | Get the number of elements in the data structure
```

Comment:
Could we specify in the comment the meaning of both elements of the returned tuple? I'm  a little confused about what the `c` value should be

Reply:
Looking at how it's used, I understand now, it's a way to transform the intermediate structure (`b`, in our case list) to the final one. I wonder if `final` is a good way to express this? Maybe `getSetWithLen` ? It's tough to name these things!

Reply:
Unfortunately it is not necessarily going to be a `Set`, it could for example be an `OSet` or even a `Map`, like in the example of scripts in the TxWits type. That's why I called it a "SetLike", cause on the wire it looks like a Set, but in Haskell it can be something else entirely.

> It's tough to name these things!

:100: 


I'll try to document the argument a little better. That should be a better solution to this problem when compared to the name of the argument anyways.